src/libraries/System.Runtime/tests/System/TimeZoneInfoTests.cs

```diff
@@ -2983,5 +3008,56 @@ private static TimeZoneInfo CreateCustomLondonTimeZone()
                 }
             });
         }
+
+        //  This helper class is used to retrieve information about installed OS languages from Windows.
+        //  Its methods returns empty when run on non-Windows platforms.
+        private static class WindowsUILanguageHelper
+        {
+            public static CultureInfo[] GetInstalledWin32CulturesWithUniqueLanguages() =>
+                GetInstalledWin32Cultures()
+                    .GroupBy(c => c.TwoLetterISOLanguageName)
+                    .Select(g => g.First())
+                    .ToArray();
+
+            public static CultureInfo[] GetInstalledWin32Cultures()
+            {
+                if (!OperatingSystem.IsWindows())
+                {
+                    return new CultureInfo[0];
+                }
+
+                var context = new EnumContext();
+                EnumUILanguagesProc proc = EnumUiLanguagesCallback;
+
+                EnumUILanguages(proc, MUI_ALL_INSTALLED_LANGUAGES | MUI_LANGUAGE_NAME, context);
+                GC.KeepAlive(proc);
+                return context.InstalledCultures.ToArray();
+            }
+
+            private static bool EnumUiLanguagesCallback(IntPtr lpUiLanguageString, EnumContext lParam)
+            {
+                var cultureName = Marshal.PtrToStringUni(lpUiLanguageString);
+                if (cultureName != null)
+                {
+                    lParam.InstalledCultures.Add(CultureInfo.GetCultureInfo(cultureName));
+                    return true;
+                }
+
+                return false;
+            }
+
+            private const uint MUI_LANGUAGE_NAME = 0x8;
+            private const uint MUI_ALL_INSTALLED_LANGUAGES = 0x20;
+
+            [DllImport("Kernel32.dll", CharSet = CharSet.Auto)]
+            private static extern bool EnumUILanguages(EnumUILanguagesProc lpUILanguageEnumProc, uint dwFlags, EnumContext lParam);
```

Comment:
Note that EnumContext here will get marshalled as `IUnknown*`. It may happen to work (I am not 100% sure), but it is very non-obvious what is going on. It is also what leads to the runtime assert since it is very unusual pattern that we apparently do not have a test coverage for.

It would be better to use the standard GCHandle or pointer here.


 It would be much more obvious

Reply:
> It would be better to use the standard GCHandle or pointer here.

We are standardizing on function pointers instead of delegates for all callbacks, so it is what I would recommend to use here as well. 

https://github.com/dotnet/runtime/blob/01b7e73cd378145264a7cb7a09365b41ed42b240/src/libraries/Common/src/Interop/Interop.Calendar.cs#L19

Reply:
Will do.  Thanks!